File: tcp_tutorial/EchoClient.java
// EchoClient.java
// Sends message to echo server EchoServer and waits for reception of echo
// Simulates a delay due to long lasting program code

import ch.aplu.tcp.*;
import ch.aplu.util.*;

public class EchoClient extends Console
  implements TcpNodeListener
{
  private final String sessionID = "echo&1&&asdf87u823";
  private final String nickname = "echoclient";
  private TcpNode tcpNode = new TcpNode();
  private boolean replied;
 
  public EchoClient()
  {
    tcpNode.addTcpNodeListener(this);
    System.out.println("Trying to connect to "
      + tcpNode.getRelay() + " using port " + tcpNode.getPort());
    tcpNode.connect(sessionID, nickname, 2, 2, 2);
    Monitor.putSleep();
    setTitle("Echo Client Ready");
    int n = 1;
    while (true)
    {
      replied = false;
      String msg = "#" + n++;
      tcpNode.sendMessage(msg);
      System.out.println("\nMessage sent: " + msg);
      System.out.println("Working now...");
      // Simulate some work
      // Set this delay time to 1000 for a 'fast' client and
      // the corresponding delay time in EchoServer to 5000 for a 'lazy' server
      TcpTools.delay(5000);  
      System.out.println("Working done.");
      System.out.println("Waiting for reply if necessary...");
      // Wait until work is finished AND got reply
      while (!replied)
        TcpTools.delay(1);
    }
  }

  public void nodeStateChanged(TcpNodeState state)
  {
    switch (state)
    {
      case DISCONNECTED:
        System.out.println("State changed: DISCONNECTED");
        break;
      case CONNECTING:
        System.out.println("State changed: CONNECTING");
        break;
      case CONNECTED:
        System.out.println("State changed: CONNECTED");
        Monitor.wakeUp();
        break;
    }
  }

  public void messageReceived(String sender, String text)
  {
    System.out.println("Message received: " + text);
    replied = true;
  }

  public void statusReceived(String text)
  {
    System.out.println("Status: " + text);
  }

  public static void main(String[] args)
  {
    new EchoClient();
  }
}

File: python/TcpEx/EchoClient.py
# EchoClient.py
# Sends message to echo server EchoServer and waits for reply

import sys
sys.path.append("./Lib/TcpJLib.jar")

from ch.aplu.tcp import TcpNode, TcpNodeState, TcpTools
from ch.aplu.util import ModelessOptionPane

def nodeStateChanged(state):
   if state == TcpNodeState.DISCONNECTED:
      print "State changed: DISCONNECTED"
   if state == TcpNodeState.CONNECTING:
       print "State changed: CONNECTING"
   if state == TcpNodeState.CONNECTED:  
      print "State changed: CONNECTED"
  
def messageReceived(sender, message):
   global replied
   print "Message received: " + message
   replied = True

def statusReceived(status):
   print "Got status: " + status

def actionPerformed(evt):
   notifyExit()
 
def notifyExit():
   global isRunning
   tcpNode.disconnect()
   mop.dispose()
   isRunning = False

sessionID = "echo&1&&asdf87u823";
nickname = "echoclient";
tcpNode = TcpNode(nodeStateChanged = nodeStateChanged, 
  messageReceived = messageReceived, statusReceived = statusReceived)
print "Trying to connect to " + tcpNode.getRelay() + \
  " using port " + str(tcpNode.getPort())
tcpNode.connect(sessionID, nickname, 2, 2, 2)
while tcpNode.getNodeState() != TcpNodeState.CONNECTED:
   TcpTools.delay(1)
print "Echo client ready"
mop = ModelessOptionPane("Client running. Shutdown?", None, "Ok", 
  actionPerformed = actionPerformed, notifyExit = notifyExit)

n = 1
isRunning = True
while isRunning:
   replied = False
   msg = "#" + str(n)
   n += 1
   tcpNode.sendMessage(msg)
   print "Message sent: " + msg
   print "Waiting for reply..."
   while not replied and isRunning:
      TcpTools.delay(1)
print "All done. Disconnecting..."
tcpNode.disconnect()  
